### PR TITLE
Refining the offline tooling for training artifact generation

### PR DESCRIPTION
**Pull request by baijumeswani**
### Description
Updates to the offline tooling to make it cleaner, and easier to use.
- Added a high level api called `generate_artifacts` that will make it easier to generate training artifacts.
- Made it mandatory for users to provide argument names that require gradient.
  - There is no easy way right now to determine which model initializer is trainable/non trainable/constant. So, we move the burden to the user. If we are able to get this information from the onnx model, this process can be made easier.
  - As a consequence, only args requiring grad are saved to the checkpoint file.
- Added documentation for most of the public facing classes.
- Updated class names to be more representative of what they are doing.
- Added `L1Loss`

### Motivation and Context
To make it easier and more intuitive to generate training artifacts for learning on the edge scenarios.

### CODE REVIEW

File: orttraining/orttraining/python/training/onnxblock/loss/__init__.py [hunk 1]
```diff
@@ -1,4 +1,4 @@
 # Copyright (c) Microsoft Corporation. All rights reserved.
 # Licensed under the MIT License.
 
-from onnxruntime.training.onnxblock.loss.loss import MSELoss, CrossEntropyLoss, BCEWithLogitsLoss, L1Loss  # noqa: F401
+from onnxruntime.training.onnxblock.loss.loss import BCEWithLogitsLoss, CrossEntropyLoss, L1Loss, MSELoss  # noqa: F401
```
Comment (justinchuby) [inline orttraining/orttraining/python/training/onnxblock/loss/__init__.py:4]: I would add `__all__` to this `__init__` file to remove the noqa.

File: orttraining/orttraining/python/training/onnxblock/optim/optim.py [hunk 2]
```diff
@@ -1,30 +1,24 @@
...
-import typing
+from typing import Optional, Tuple
 
 import onnx
 
-import onnxruntime.training.onnxblock._graph_utils as graph_utils
-import onnxruntime.training.onnxblock.building_blocks as building_blocks
-import onnxruntime.training.onnxblock.model as model
-import onnxruntime.training.onnxblock.model_accessor as accessor
+import onnxruntime.training.onnxblock._graph_utils as _graph_utils
```
Comment (justinchuby) [inline orttraining/orttraining/python/training/onnxblock/optim/optim.py:8]:
```suggestion
from onnxruntime.training.onnxblock import _graph_utils
```

consider this. here and elsewhere?

File: orttraining/orttraining/python/training/onnxblock/optim/__init__.py [hunk 3]
```diff
@@ -1,5 +1,4 @@
 # Copyright (c) Microsoft Corporation. All rights reserved.
 # Licensed under the MIT License.
-# __init__.py
 
-from .optim import AdamW, ClipGradNorm  # noqa: F401
+from onnxruntime.training.onnxblock.optim.optim import AdamW, ClipGradNorm  # noqa: F401
```
Comment (justinchuby) [inline orttraining/orttraining/python/training/onnxblock/optim/__init__.py:4]: Nice absolute import!

File: orttraining/orttraining/python/training/onnxblock/model_accessor.py [hunk 4]
```diff
@@ -24,43 +28,90 @@ def model(self):
...
 
-    # If the user did not provide a model, then assume that they want to build from scratch.
-    # It is the duty of the caller to fill the model however they deem fit.
-    if model is None:
-        model = onnx.ModelProto()
+    model_clone = copy.deepcopy(model)
+
+    if model_clone is None:
+        raise RuntimeError(
+            "Base onnx model cannot be None. Please use onnxblock.empty_base if you would like to build a \
```
Comment (justinchuby) [inline orttraining/orttraining/python/training/onnxblock/model_accessor.py:65]:
```suggestion
            "Base onnx model cannot be None. Please use onnxblock.empty_base if you would like to build a "
```

File: orttraining/orttraining/python/training/onnxblock/model_accessor.py [hunk 5]
```diff
@@ -24,43 +28,90 @@ def model(self):
...
+                model from scratch."
```
Comment (justinchuby) [inline orttraining/orttraining/python/training/onnxblock/model_accessor.py:66]:
```suggestion
                "model from scratch."
```

File: orttraining/orttraining/python/training/onnxblock/model_accessor.py [hunk 6]
```diff
@@ -24,43 +28,90 @@ def model(self):
...
+    >>> with onnxblock.base(model) as model_handle:
+    >>>     # manipulate the model using blocks
+    >>>     ...
+    >>>     # get the modified model
+    >>>     model = model_handle.model
+
+    In this example, base will register the given input model to be manipulated by the onnx blocks.
+
+    Args:
+        model (onnx.ModelProto): The base model to be manipulated by the onnx blocks.
```
Comment (justinchuby) [inline orttraining/orttraining/python/training/onnxblock/model_accessor.py:52]:
```suggestion
        model: The base model to be manipulated by the onnx blocks.
```

consider adding type to the signature?

File: orttraining/orttraining/python/training/onnxblock/model_accessor.py [hunk 7]
```diff
@@ -1,18 +1,22 @@
...
+from typing import Optional
 
 import onnx
 
 
 class ModelAccessor:
-    """This class stores the onnx model that is manipulated by the onnx blocks."""
+    """This class stores the onnx model that is manipulated by the onnx blocks.
+
+    Attributes:
```
Comment (justinchuby) [inline orttraining/orttraining/python/training/onnxblock/model_accessor.py:14]: Consider annotating the return type of model()

File: orttraining/orttraining/python/training/onnxblock/blocks.py [hunk 8]
```diff
@@ -302,14 +310,52 @@ def build(self, clip_input_name):
...
+        Args:
+            like (str): The name of the graph input/output to copy the type and shape from
+        """
+        super().__init__()
+
+        self._like = like
+
+    def build(self, input_name: Optional[str] = None):
+        cloned_input = None
+        with contextlib.suppress(LookupError):
```
Comment (justinchuby) [inline orttraining/orttraining/python/training/onnxblock/blocks.py:348]: Comment on why it’s suppressed?

File: orttraining/orttraining/training_api/checkpoint.cc [hunk 9]
```diff
@@ -535,8 +535,8 @@ Status OrtLoadInternal(const PathString& checkpoint_path,
   for (auto& init : *(model_proto.mutable_graph()->mutable_initializer())) {
     ORT_ENFORCE(init.has_name(), "An initializer should have a name.");
     auto it = param_tensor_protos.find(init.name());
-    ORT_ENFORCE(it != param_tensor_protos.end(),
-                "The initializer name was not found in the checkpoint file loaded.");
+    if (it == param_tensor_protos.end())
```
Comment (pengwa) [inline orttraining/orttraining/training_api/checkpoint.cc:538]: what's the reason we make the change?

File: orttraining/orttraining/python/training/onnxblock/checkpoint_utils.py [hunk 10]
```diff
@@ -1,26 +1,30 @@
...
 
     if parameters is None:
         raise RuntimeError("No checkpoint parameters provided.")
 
-    # TODO: use Parameter class to pass information to backend
-    # Serialize the parameters and save the checkpoint
     trainable_params, non_trainable_params = parameters
     trainable_params = [param.SerializeToString() for param in trainable_params]
     non_trainable_params = [param.SerializeToString() for param in non_trainable_params]
     _internal_save_checkpoint(trainable_params, non_trainable_params, path_to_checkpoint)
```
Comment (pengwa) [inline orttraining/orttraining/python/training/onnxblock/checkpoint_utils.py:24]: in the PR description "- As a consequence, only args requiring grad are saved to the checkpoint file."

is that valid?

File: orttraining/orttraining/python/training/artifacts.py [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+
+
+def generate_artifacts(
+    model: onnx.ModelProto,
+    requires_grad: Optional[List[str]] = None,
+    loss: Optional[Union[LossType, onnxblock.Block]] = None,
+    optimizer: Optional[OptimType] = None,
+    artifact_directory: Optional[Union[str, os.PathLike]] = None,
+    **extra_options,
+) -> None:
```
Comment (pengwa) [inline orttraining/orttraining/python/training/artifacts.py:42]: shall we have comment to make it clear what the artifacts contain?

File: orttraining/orttraining/python/training/artifacts.py [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+    L1Loss = 4
+
+
+class OptimType(Enum):
+    """Enum to represent the optimizers supported by ORT
+
+    To be used with the `optimizer` parameter of `generate_artifacts` function.
+    """
+
+    AdamW = 1
```
Comment (pengwa) [inline orttraining/orttraining/python/training/artifacts.py:32]: this reminds me I added SGD optimizer but did not add related changes in offline processing yet. :) 
let me add an item track it

**Review by pengwa — COMMENTED** [review]
Went through it quickly, left few comments.

File: orttraining/orttraining/python/training/onnxblock/__init__.py [hunk 13]
```diff
@@ -1,11 +1,12 @@
...
 
-"""Offline tooling for generating files needed for ort training apis."""
+"""Offline tooling for generating files needed by ort training apis."""
 
-from . import loss, optim  # noqa: F401
-from .building_blocks import Block  # noqa: F401
-from .checkpoint_utils import load_checkpoint_to_model, save_checkpoint  # noqa: F401
-from .model import Model, TrainingModel  # noqa: F401
-from .model_accessor import onnx_model  # noqa: F401
+import onnxruntime.training.onnxblock.blocks as blocks  # noqa: F401
```
[the remark this answers is not in the document]
Comment (baijumeswani) [reply orttraining/orttraining/python/training/onnxblock/__init__.py:6]: Typically, the recommendation is to use absolute imports. The package name is `onnxruntime`. 
It is safe to assume that these modeuls will not be used outside the context of onnxruntime package. So, this should always be valid.

File: orttraining/orttraining/python/training/onnxblock/optim/optim.py (hunk 2 above)
Comment (baijumeswani) [reply orttraining/orttraining/python/training/onnxblock/optim/optim.py:8]: Is there an advantage in doing it this way.

My thought process is to distinguish imports from __init__.py and local files with this way of importing and is purely subjective.

I don't know if there is any advantage of doing one way over the other.

File: orttraining/orttraining/training_api/checkpoint.cc (hunk 9 above)
Comment (baijumeswani) [reply orttraining/orttraining/training_api/checkpoint.cc:538]: I think this was an error.

Imagine there is a checkpoint file that has only a partial list of initializers to load. The earlier code will result in a failure because the initializer in the model are not there in the checkpoint file.

We should only load the initializers in the checkpoint file on the model, without caring for the existence of all the model initializers in the checkpoint.

Comment (baijumeswani) [reply orttraining/orttraining/training_api/checkpoint.cc:538]: Maybe we can expose a `strict` bool flag that enforces/relaxes this rule.

File: orttraining/orttraining/python/training/onnxblock/optim/optim.py (hunk 2 above)
Comment (justinchuby) [reply orttraining/orttraining/python/training/onnxblock/optim/optim.py:8]: Just to avoid repetition. NBD. I think we should use as only when renaming.

File: orttraining/orttraining/python/training/onnxblock/blocks.py [hunk 14]
```diff
@@ -302,14 +310,52 @@ def build(self, clip_input_name):
...
+            cloned_input = copy.deepcopy(_graph_utils.get_input_from_input_name(self.base, self._like))
+
+        if cloned_input is None:
+            with contextlib.suppress(LookupError):
+                cloned_input = copy.deepcopy(_graph_utils.get_output_from_output_name(self.base, self._like))
+
+        if cloned_input is None:
```
[the remark this answers is not in the document]
Comment (baijumeswani) [reply orttraining/orttraining/python/training/onnxblock/blocks.py:355]: This block is creating an input like either the input or output associated with the given name.
`InputLike` creates a `ValueInfoProto` and appends to model.graph.input.

Similarly, `OutputsLike` would append to model.graph.output.

But both blocks would look into existing inputs/outputs of the graph. Would the name `InputLikeValueProtoInfo` make the intent clearer?

File: orttraining/orttraining/python/training/onnxblock/optim/optim.py (hunk 2 above)
Comment (baijumeswani) [reply orttraining/orttraining/python/training/onnxblock/optim/optim.py:8]: I'll let this be as is since I would like to avoid having to type out `onnxruntime.training.onnxblock._graph_utils` over and over again.

If you feel strongly about this, I will go ahead and make the change.

Comment (justinchuby) [reply orttraining/orttraining/python/training/onnxblock/optim/optim.py:8]: Just making sure, I think `from onnxruntime.training.onnxblock import _graph_utils` will save you from typing the whole thing too?

Comment (baijumeswani) [reply orttraining/orttraining/python/training/onnxblock/optim/optim.py:8]: Spoke to Justin offline. Will keep this as is for now. Thanks, @justinchuby

File: orttraining/orttraining/python/training/onnxblock/blocks.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (pengwa) [reply orttraining/orttraining/python/training/onnxblock/blocks.py:357]: I see, I think it is fine to keep it. Thanks for the explanation.

File: orttraining/orttraining/python/training/onnxblock/checkpoint_utils.py (hunk 10 above)
Comment (pengwa) [reply orttraining/orttraining/python/training/onnxblock/checkpoint_utils.py:29]: OK. May I know why "non_trainable_params" is empty?  and maybe non trainable params should also be specfied by users?

File: orttraining/orttraining/training_api/checkpoint.cc [hunk 15]
```diff
@@ -535,16 +536,20 @@ Status OrtLoadInternal(const PathString& checkpoint_path,
...
+    }
     init = it->second;
   }
 
   return Status::OK();
 }
 
 Status OrtLoadInternal(const PathString& checkpoint_path, CheckpointState& state) {
-  ORT_ENFORCE(Env::Default().FolderExists(checkpoint_path), "Checkpoint folder not exit");
+  ORT_ENFORCE(Env::Default().FolderExists(checkpoint_path), "Checkpoint folder does not exist.");
```
Comment (pengwa) [inline orttraining/orttraining/training_api/checkpoint.cc:552]: :)

File: orttraining/orttraining/training_api/checkpoint.cc (hunk 9 above)
Comment (pengwa) [reply orttraining/orttraining/training_api/checkpoint.cc]: Are the constant initializers stored in model proto now? and previously they is stored as non-trainable params, Am I right?

File: orttraining/orttraining/python/training/artifacts.py [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+
+    training_model = None
+    eval_model = None
+    model_params = None
+    with onnxblock.base(model):
+        _ = training_block(*[output.name for output in model.graph.output])
+        training_model, eval_model = training_block.to_model_proto()
+        model_params = training_block.parameters()
+
+    if artifact_directory is None:
```
[the remark this answers is not in the document]
Comment (pengwa) [reply orttraining/orttraining/python/training/artifacts.py:116]: Not need to make so complex so far, this is fine.

File: orttraining/orttraining/training_api/checkpoint.cc (hunk 9 above)
Comment (baijumeswani) [reply orttraining/orttraining/training_api/checkpoint.cc]: There are three types of initializers:
1. Trainable model parameters.
2. Non trainable model parameters.
3. Constant model initializers.

Previously, 1 and 2 were in the checkpoint file, where as 3 was in model proto.

Now, only 1 is in checkpoint file, and 2 and 3 are in model proto.

File: orttraining/orttraining/python/training/onnxblock/checkpoint_utils.py (hunk 10 above)
Comment (baijumeswani) [reply orttraining/orttraining/python/training/onnxblock/checkpoint_utils.py:29]: We don't know how to distinguish non trainable params from constant initializers. So, we make both these as part of model proto. Which is why non trainable params are empty here.

File: orttraining/orttraining/python/training/onnxblock/README.md [hunk 17]
```diff
@@ -11,95 +11,102 @@ The ort training APIs need the following files for performing training
...
+from onnxruntime.training import artifacts
 
-# Define how the training model should be constructed.
-# In this example, the training model is constructed by stacking MSELoss
-# blocks on top of the inference model (with two outputs) and taking a
-# weighted average of the two losses to compute the final loss.
-# And the onnxblock tooling automatically takes care of building the
-# training graph for this model.
-class MyTrainingModel(onnxblock.TrainingModel):
+# Define a custom loss block that takes in two inputs
```
Comment (xadupre) [inline orttraining/orttraining/python/training/onnxblock/README.md:73]: Why not using torch to write the custom loss?

Comment (baijumeswani) [reply orttraining/orttraining/python/training/onnxblock/README.md:73]: This will require users to have torch in their offline tooling ecosystem.

While this could be true for many cases, it could also not be a dependency that the user would like to take.

For example, if the user begins with a model from model zoo, they would probably not like to take a dependency on torch for defining a custom loss.

But, as we spoke earlier, I want to keep this as a possibility. So, we may open up the input argument for custom loss to be a custom torch loss function. But I will prioritize this in a separate PR.

File: orttraining/orttraining/python/training/onnxblock/checkpoint_utils.py (hunk 10 above)
Comment (baijumeswani) [reply orttraining/orttraining/python/training/onnxblock/checkpoint_utils.py:29]: I made the change so users must pass in the frozen params in case they know this information.

File: orttraining/orttraining/python/training/artifacts.py [hunk 18]
```diff
@@ -105,6 +107,10 @@ def build(self, *inputs_to_loss):
         for arg in requires_grad:
             training_block.requires_grad(arg)
 
+    if frozen_params is not None:
+        for arg in frozen_params:
+            training_block.requires_grad(arg, False)
+
```
Comment (pengwa) [inline orttraining/orttraining/python/training/artifacts.py:113]: nit: maybe we should add a check when in frozen_prams loops in case users pass a weight name in both requires_grad and frozen_params.

**Review by pengwa — DISMISSED** [review]

File: orttraining/orttraining/python/training/artifacts.py (hunk 18 above)
Comment (baijumeswani) [reply orttraining/orttraining/python/training/artifacts.py:113]: There is a check inside training Blick. So it should be OK for user to add to both.

Comment (baijumeswani) [reply orttraining/orttraining/python/training/artifacts.py:113]: But I think having a check here makes sense. Will add it.

**Review by pengwa — APPROVED** [review]

**Review by pengwa — APPROVED** [review]
